### PR TITLE
internal/dag: scan pem for secret key

### LINKED ISSUE #1702: TLS keys containing other pem blocks proceeding the private key are considered invalid
**What steps did you take and what happened:**
Created a secret whose `EC PRIVATE KEY` contains `EC PARAMETERS`.  The function `validatePrivateKey` locates the `EC PARAMETERS ` pem.Block, attempts to validate it as a private key, and rejects the secret as valid.

**What did you expect to happen:**
I expected the secret to be accepted as valid :)

**Anything else you would like to add:**
The private key pem struct is something like this

```
-----BEGIN EC PARAMETERS-----
-----END EC PARAMETERS-----
-----BEGIN EC PRIVATE KEY-----
-----END EC PRIVATE KEY-----
```

I did scan, but didn't find a similar issue.

**Environment:**

- Contour version: v0.15.1 and master
- Kubernetes version: (use `kubectl version`): N/A
- Kubernetes installer & version: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from `/etc/os-release`): N/A

### PR DESCRIPTION
**Pull request by mattalberts**
Walk each PEM block to identify the PRIVATE KEY

closes #1702

### CODE REVIEW

File: internal/dag/secret.go [hunk 1]
```diff
@@ -59,9 +60,9 @@ func validateCertificate(data []byte) error {
 }
 
 func validatePrivateKey(data []byte) error {
-	key, err := decodePEM(data)
+	key, err := decodePEMPrivateKey(data)
 	if err != nil {
-		return nil
+		return err
```
Comment (mattalberts) [inline internal/dag/secret.go]: I wasn't sure why err was not previously returned (failure to parse the PEM should be an error?)

File: internal/dag/secret.go [hunk 2]
```diff
@@ -82,3 +83,21 @@ func decodePEM(in []byte) (*pem.Block, error) {
...
+		}
+		if strings.HasSuffix(block.Type, "PRIVATE KEY") {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil, errors.New("failed to locate private key")
+	}
+	return block, nil
```
Comment (mattalberts) [inline internal/dag/secret.go]: there is an opportunity to be ""smarter" with the outbound block and switch on type to identify the parsing method.

File: internal/dag/secret.go [hunk 3]
```diff
@@ -82,3 +83,16 @@ func decodePEM(in []byte) (*pem.Block, error) {
...
 }
+
+func decodePEMPrivateKey(in []byte) (*pem.Block, error) {
+	var block *pem.Block
+	for len(in) > 0 {
+		if block, in = pem.Decode(in); block == nil {
+			return nil, errors.New("failed to parse PEM")
+		}
+		if strings.HasSuffix(block.Type, "PRIVATE KEY") {
+			return block, nil
```
Comment (mattalberts) [inline internal/dag/secret.go]: there is an opportunity to be ""smarter" with the outbound block and switch on type to identify the appropriate parsing method.

**Comment by jpeach** [comment]
I can review tomorrow, if no-one else gets there first :)

**Comment by mattalberts** [comment]
Thanks!  I very much drafted this change set in the "backport to v0.15.1" mindset.   That said, the certificate parsing/validation has a similar flaw (to less detriment).  It will accept any cert whose first pem block is a valid cert.  I'll do mock up an alternate that might be better for mainline

**Comment by youngnick** [comment]
I think that #1707 is the correct approach here, so I'll close this one out. Thanks for all your work @mattalberts!

**Pull request closed without merging**